mllib/src/main/scala/org/apache/spark/mllib/stat/distribution/MultivariateGaussian.scala

```diff
@@ -126,11 +120,11 @@ class MultivariateGaussian @Since("1.3.0") (
       // log(pseudo-determinant) is sum of the logs of all non-zero singular values
       val logPseudoDetSigma = d.activeValuesIterator.filter(_ > tol).map(math.log).sum
 
-      // calculate the root-pseudo-inverse of the diagonal matrix of singular values
-      // by inverting the square root of all non-zero values
-      val pinvS = diag(new DBV(d.map(v => if (v > tol) math.sqrt(1.0 / v) else 0.0).toArray))
+      // calculate the pseudo-inverse of the diagonal matrix of singular values
+      // by inverting the non-zero values
+      val pinvS = diag(new DBV(d.map(v => if (v > tol) 1.0 / v else 0.0).toArray))
 
-      (pinvS * u, -0.5 * (mu.size * math.log(2.0 * math.Pi) + logPseudoDetSigma))
+      (u * pinvS * u.t, -0.5 * (mu.size * math.log(2.0 * math.Pi) + logPseudoDetSigma))
```

Comment:
There is an extra matrix-matrix multiplication in `u * pinvS * u.t`. I think the bug is in line 133, where we should use `pinvS * u.t` instead of `pinvS * u`. Could you check this solution? Some comments need updates too.